src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs

```diff
@@ -0,0 +1,196 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Roslyn.Utilities;
+using AsyncCompletionData = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
+using VSCompletionItem = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data.CompletionItem;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.AsyncCompletion
+{
+    internal sealed class CommitManager : IAsyncCompletionCommitManager
+    {
+        private static readonly IEnumerable<char> s_commitChars = ImmutableArray.Create(
+            ' ', '{', '}', '[', ']', '(', ')', '.', ',', ':',
+            ';', '+', '-', '*', '/', '%', '&', '|', '^', '!',
+            '~', '=', '<', '>', '?', '@', '#', '\'', '\"', '\\');
```

Comment:
i don't know what these are for... or why this is defined here.  Commit characters are language specific.

Reply:
This is an attempt to list possible characters for all languages supported by Roslyn. If we will not be able to maintain them, we:
1. Will always return true for ShouldCommitCompletion
2. Will ask the editor to depreciate PotentialCommitCharacters on their side.

Attn: @AmadeusW

---
In reply to: [222519349](https://github.com/dotnet/roslyn/pull/29016#discussion_r222519349) [](ancestors = 222519349)

Reply:
> This is an attempt to list possible characters for all languages supported by Roslyn. If we will not be able to maintain them, we:

Understood. However, this is a dynamic entrypoint in roslyn, and any languages *or 3rd party* can specify *any* characters they want.  Those characters are dynamic to the level of individual items themselves.  I don't know how could enumerate a fixed set *except* to make it the entire set of characters :)  In which case, we should always return 'true'.

Reply:
Actually, i'm slightly wrong here.  Roslyn goes the *inverse* route.  It hardcodes the list of *non* commit characters.  That list is exactly:

```c#
            return char.IsLetter(args.TypedChar)
                || char.IsNumber(args.TypedChar)
                || args.TypedChar == '_';
```

So the right way to implement ths is to do:

```c#
            return !char.IsLetter(args.TypedChar)
                && !char.IsNumber(args.TypedChar)
                && args.TypedChar != '_';
```

Reply:
OK. Let us try with the current approach. If it would not work, we know what to do.

---
In reply to: [222868237](https://github.com/dotnet/roslyn/pull/29016#discussion_r222868237) [](ancestors = 222868237)

Reply:
Makes sense. Thank you!

---
In reply to: [222868454](https://github.com/dotnet/roslyn/pull/29016#discussion_r222868454) [](ancestors = 222868454)